File: agent/capabilities.py
"""Capability catalog — the agent's safety cage, serialized from enforcement constants.

``build_capabilities()`` assembles the DTO returned by ``GET /capabilities``
(Task 3). Every field it returns is derived from the SAME constants that
the enforcement code imports — never from hand-written documentation that
could silently drift. The coupling is verified by the drift-pin tests in
``tests/unit/test_capabilities.py``:

- ``test_tool_descriptions_cover_exactly_the_tool_registry`` — every key
  in ``TOOL_REGISTRY`` has a description and no stale entries exist.
- ``test_worker_descriptions_cover_exactly_the_worker_registry`` — same
  for ``WORKER_REGISTRY``.
- ``test_rule_categories_cover_exactly_the_rule_descriptions`` — every
  rule in ``RULE_DESCRIPTIONS`` has a category assignment.
- ``test_every_approval_gated_action_has_a_human_gate`` — every
  ``ACTION_REGISTRY`` entry with ``requires_approval=True`` appears in
  ``HUMAN_GATES``.
- ``test_chat_only_coherence_with_main`` — ``observation_kind == "none"``
  in the manifests agrees with ``CHAT_ONLY_WORKLOAD_NAMES`` in main.py.

Import direction: this module imports FROM ``agent.workloads.registry``,
``agent.workloads.spec``, ``agent.fanout``, and
``driftscribe_lib.iac_plan_denylist``. It MUST NOT import ``agent.main``
AT MODULE LOAD — main imports us, so a top-level import would cycle.
``build_capabilities()`` does take a single deliberate function-scope
import of ``agent.main.AUTONOMOUS_TRIGGER_WORKLOADS`` at call time (when
main is already fully loaded), mirroring the test side's lazy
``from agent.main import CHAT_ONLY_WORKLOAD_NAMES`` — see the comment at
the import site for why the autonomy signal is owned in main.
"""
from __future__ import annotations

import dataclasses
from types import MappingProxyType
from typing import Final, Mapping, get_args

from driftscribe_lib.iac_plan_denylist import RULE_DESCRIPTIONS, ADOPTABLE_RESOURCE_TYPES

from agent.fanout import MUTATION_TOOL_NAMES
from agent.workloads.registry import ACTION_REGISTRY
from agent.workloads.registry import load_workload_spec
from agent.workloads.spec import WorkloadSpec


# --------------------------------------------------------------------------- #
# WORKLOAD_NAMES — closed set, derived from the Literal in WorkloadSpec.name
# --------------------------------------------------------------------------- #

WORKLOAD_NAMES: Final[tuple[str, ...]] = get_args(
    WorkloadSpec.model_fields["name"].annotation
)
"""The ordered workload names, derived from the ``WorkloadSpec.name`` Literal.

The Literal IS the enumeration — we never hand-list workload names here.
Order follows the Literal declaration in ``agent/workloads/spec.py``
(currently ``("drift", "upgrade", "explore", "provision")``).

Pinned by ``test_build_capabilities_shape``: ``[w["name"] for w in dto["workloads"]]``
must equal ``list(WORKLOAD_NAMES)``.
"""


# --------------------------------------------------------------------------- #
# TOOL_DESCRIPTIONS — one entry per TOOL_REGISTRY key
# --------------------------------------------------------------------------- #

TOOL_DESCRIPTIONS: Final[Mapping[str, str]] = MappingProxyType({
    "drift_read_live_env": (
        "Reads the live Cloud Run environment: deployed image, revision, "
        "environment variables, and service configuration."
    ),
    "read_project_inventory": (
        "Reads a read-only whole-project asset inventory via the infra-reader "
        "worker (Cloud Asset Viewer only, no write access)."
    ),
    "drift_patch_docs": (
        "Updates the ops-contract documentation to record the current observed "
        "state after a drift detection run."
    ),
    "drift_propose_rollback": (
        "Proposes a rollback; never executes one. It creates an approval that "
        "waits for an operator."
    ),
    "notify": (
        "Sends a notification via the notifier worker (counted as write-capable "
        "because it rides a sending credential)."
    ),
    "load_contract": (
        "Loads the ops-contract YAML (the declarative ground truth) so the agent "
        "can compare it with observed state."
    ),
    "search_recent_prs": (
        "Searches the target repo's recent pull requests (counted as write-capable "
        "because it rides a repo credential)."
    ),
    "load_iac_plan": (
        "Reads the latest verified plan artifact for a pending infrastructure "
        "PR and summarizes it in plain language. Read-only: cannot approve, "
        "reject, or apply anything."
    ),
    "upgrade_read_dependencies": (
        "Reads the target repo's dependency lockfile to identify outdated packages."
    ),
    "upgrade_propose_pr": (
        "Opens a dependency-upgrade pull request in the target repo."
    ),
    "upgrade_close_pr": (
        "Closes an upgrade PR this agent opened, only when it is safe to do so "
        "(driftscribe label, upgrade/ branch, correct base)."
    ),
    "upgrade_merge_pr": (
        "Merges an upgrade PR this agent opened, only after CI is green on the "
        "exact head commit. Fails closed."
    ),
    "search_developer_docs": (
        "Searches the developer knowledge base for documentation relevant to the "
        "current task."
    ),
    "retrieve_developer_doc": (
        "Retrieves a specific document from the developer knowledge base by ID."
    ),
    "provision_open_infra_pr": (
        "Authors OpenTofu files under iac/ and opens ONE pull request. Never "
        "applies anything; applying happens only through the gated "
        "approve-then-apply pipeline."
    ),
    "provision_propose_adoption": (
        "Adopt an existing resource into IaC management via a zero-change "
        "import PR. Renders the config deterministically; cannot modify live "
        "infrastructure."
    ),
    "get_session_state": (
        "Reserved; not implemented. No workload can use it."
    ),
    "set_session_state": (
        "Reserved; not implemented. No workload can use it."
    ),
})
"""Operator-facing description of every tool in ``TOOL_REGISTRY``.

Keyed by the EXACT symbolic names in ``TOOL_REGISTRY`` — the drift-pin test
``test_tool_descriptions_cover_exactly_the_tool_registry`` in
``tests/unit/test_capabilities.py`` asserts set equality so a new tool cannot
ship without a description and a deleted tool cannot leave a stale entry.

Descriptions for tools in ``MUTATION_TOOL_NAMES`` that are there for credential
containment rather than direct mutation (``notify``, ``search_recent_prs``) must
say so explicitly — the ``write_capable`` field in the DTO carries the
``MUTATION_TOOL_NAMES`` membership, but the description is where the nuance lives.
"""


# --------------------------------------------------------------------------- #
# WORKER_DESCRIPTIONS — one entry per WORKER_REGISTRY key
# --------------------------------------------------------------------------- #

WORKER_DESCRIPTIONS: Final[Mapping[str, str]] = MappingProxyType({
    "drift_reader": (
        "Reads the live Cloud Run service state for drift detection. "
        "Read-only by the scope of calls it makes."
    ),
    "drift_docs": (
        "Patches the ops-contract documentation to record observed state."
    ),
    "drift_rollback": (
        "Executes a Cloud Run rollback to a previous revision. Refuses "
        "anything without a valid operator approval token."
    ),
    "infra_reader": (
        "Reads the whole-project GCP asset inventory. Read-only by IAM "
        "(asset viewer only)."
    ),
    "notifier": (
        "Sends notifications (e.g. Slack or webhook). Carries a sending "
        "credential."
    ),
    "upgrade_reader": (
        "Reads the target repo's dependency lockfile. Read-only by the "
        "scope of calls it makes."
    ),
    "upgrade_docs": (
        "Opens and manages upgrade pull requests in the target repo."
    ),
    "tofu_editor": (
        "Writes iac/-only files and opens PRs; never touches live "
        "infrastructure."
    ),
})
"""Operator-facing description of every worker in ``WORKER_REGISTRY``.

Keyed by the EXACT symbolic names in ``WORKER_REGISTRY`` — the drift-pin test
``test_worker_descriptions_cover_exactly_the_worker_registry`` asserts set
equality. Required nuances: ``infra_reader`` must mention read-only IAM;
``tofu_editor`` must clarify it only writes files and opens PRs; ``drift_rollback``
must mention the operator approval token requirement.
"""


# --------------------------------------------------------------------------- #
# CATEGORY_ORDER and RULE_CATEGORIES — denylist rule taxonomy
# --------------------------------------------------------------------------- #

ADOPTABLE_TYPE_LABELS: Final[Mapping[str, str]] = MappingProxyType({
    "google_storage_bucket":      "Cloud Storage bucket",
    "google_pubsub_topic":        "Pub/Sub topic",
    "google_pubsub_subscription": "Pub/Sub subscription",
    "google_cloud_run_v2_service": "Cloud Run service",
})
"""Human-readable label for every adoptable resource type in ``ADOPTABLE_RESOURCE_TYPES``.

Drift-pinned by ``test_adoptable_type_labels_cover_exactly_the_allowlist``:
``set(ADOPTABLE_TYPE_LABELS) == set(ADOPTABLE_RESOURCE_TYPES)``.
"""


CATEGORY_ORDER: Final[tuple[str, ...]] = (
    "control-plane",
    "service-managed",
    "iam",
    "global-v1",
    "structural",
)
"""Rendering order for denylist rule categories (anxiety-first).

Used both to sort rules in the DTO (``(CATEGORY_ORDER.index(cat), rule_id)``)
and to group them in the frontend. Verified by ``test_build_capabilities_shape``
which asserts full sort stability.
"""

RULE_CATEGORIES: Final[Mapping[str, str]] = MappingProxyType({
    # Structural (malformed plan rejected outright):
    "plan-json-unparseable":              "structural",
    "plan-json-missing-resource-changes": "structural",
    "plan-json-malformed-change":         "structural",
    # Control-plane (DriftScribe's own infrastructure):
    "control-plane-service":  "control-plane",
    "control-plane-sa":       "control-plane",
    "control-plane-bucket":   "control-plane",
    "control-plane-secret":   "control-plane",
    "control-plane-kms":      "control-plane",
    # Service-managed (buckets other Google services auto-create):
    "service-managed-bucket": "service-managed",
    # IAM (access-control changes):
    "wif-config-change":          "iam",
    "iam-change-forbidden-v1":    "iam",
    # Global v1 floors (structural action bans + conditional admit rules):
    "import-with-changes-forbidden-v1": "global-v1",
    "import-type-not-adoptable-v1":     "global-v1",
    "import-mixed-plan-forbidden-v1":   "global-v1",
    "import-batch-forbidden-v1":        "global-v1",
    "delete-action-forbidden-v1":       "global-v1",
    "forget-action-forbidden-v1":       "global-v1",
    "replace-action-forbidden-v1":      "global-v1",
    "unknown-action-forbidden-v1":      "global-v1",
})
"""Category assignment for every rule ID in ``RULE_DESCRIPTIONS``.

Keyed by the EXACT rule IDs from ``RULE_DESCRIPTIONS`` — the drift-pin test
``test_rule_categories_cover_exactly_the_rule_descriptions`` asserts:
- ``set(RULE_CATEGORIES) == set(RULE_DESCRIPTIONS)`` (no stale/missing entries)
- ``set(RULE_CATEGORIES.values()) <= set(CATEGORY_ORDER)`` (no unknown categories)
"""


# --------------------------------------------------------------------------- #
# HUMAN_GATES — approval gates that require operator action
# --------------------------------------------------------------------------- #

# Inner dicts are frozen with MappingProxyType — same in-place-mutation
# protection convention as TOOL_REGISTRY / WORKER_REGISTRY / ACTION_REGISTRY
# (a caller holding a reference cannot poison later build_capabilities() calls).
HUMAN_GATES: Final[tuple[Mapping[str, str], ...]] = (
    MappingProxyType({
        "id": "iac_apply",
        "title": "IaC plan apply",
        "description": (
            "Before the apply worker runs ``tofu apply``, an operator must "
            "approve the exact stored plan via the approval page. The approval "
            "is bound to the specific plan by a plan-bound HMAC with a signed "
            "expiry window. Approving one plan cannot approve another."
        ),
        "route": "/iac-approvals/{pr_number}",
        "method": "POST",
    }),
    MappingProxyType({
        "id": "rollback",
        "title": "Rollback",
        "description": (
            "The rollback worker requires a valid operator approval token before "
            "it will execute any Cloud Run rollback. The approval is single-use "
            "with a 15-minute TTL and bound to the specific rollback request by "
            "HMAC. The worker re-verifies the token at execution time."
        ),
        "route": "/approvals/{approval_id}",
        "method": "POST",
    }),
)
"""The two human approval gates, in DTO-ready dict form.

``method`` is included so the frontend can pin the gate to the mutating POST
(not the GET form page). The drift-pin test
``test_every_approval_gated_action_has_a_human_gate`` verifies that every
``ACTION_REGISTRY`` entry with ``requires_approval=True`` has a corresponding
gate id here.

Gate descriptions must mention their specific approval mechanism:
- iac_apply: plan-bound HMAC + signed expiry window
- rollback: single-use + 15-minute TTL + worker-side re-verification
"""


# --------------------------------------------------------------------------- #
# build_capabilities() — assembles the DTO
# --------------------------------------------------------------------------- #

def build_capabilities() -> dict:
    """Assemble the capabilities DTO from the enforcement constants.

    Returns a plain ``dict`` / ``list`` tree (JSON-serializable by
    ``json.dumps`` without a custom encoder). Never reads worker-URL
    env vars — it uses :func:`agent.workloads.registry.load_workload_spec`
    (env-free symbol validation only) so it works in every deploy
    environment.

    Does not cache; the workload YAML is re-parsed on every call (static
    per deploy, four small files — deliberate).

    Ordering is deterministic:
    - Workloads in ``WORKLOAD_NAMES`` order (Literal declaration order).
    - Tools / workers / actions in manifest declaration order.
    - Denylist rules sorted by ``(CATEGORY_ORDER.index(category), rule_id)``.

    Pinned by ``tests/unit/test_capabilities.py::test_build_capabilities_shape``
    and ``test_build_capabilities_is_json_serializable_and_env_free``.
    """
    # The "autonomous" signal is the explicit trigger set owned in
    # agent.main (NOT derived from observation_kind, which is intent not a
    # wired trigger — only drift actually fires on its own). Imported lazily
    # here, mirroring tests/unit/test_capabilities.py::
    # test_chat_only_coherence_with_main, so this module never imports the
    # heavy FastAPI app at load time (no cycle: main imports capabilities).
    from agent.main import AUTONOMOUS_TRIGGER_WORKLOADS

    workloads = []
    for name in WORKLOAD_NAMES:
        spec = load_workload_spec(name)
        tools = [
            {
                "name": tool_name,
                "description": TOOL_DESCRIPTIONS[tool_name],
                "write_capable": tool_name in MUTATION_TOOL_NAMES,
            }
            for tool_name in spec.enabled_tool_names
        ]
        workers = [
            {
                "name": worker_name,
                "description": WORKER_DESCRIPTIONS[worker_name],
            }
            for worker_name in spec.worker_names
        ]
        actions = [
            dataclasses.asdict(ACTION_REGISTRY[action_name])
            for action_name in spec.action_names
        ]
        workloads.append({
            "name": spec.name,
            "display_name": spec.display_name,
            "descriptor": spec.descriptor,
            "description": spec.description,
            "autonomous": spec.name in AUTONOMOUS_TRIGGER_WORKLOADS,
            "tools": tools,
            "workers": workers,
            "actions": actions,
        })

    rules = sorted(
        [
            {
                "id": rule_id,
                "description": RULE_DESCRIPTIONS[rule_id],
                "category": RULE_CATEGORIES[rule_id],
            }
            for rule_id in RULE_DESCRIPTIONS
        ],
        key=lambda r: (CATEGORY_ORDER.index(r["category"]), r["id"]),
    )

    return {
        "version": 1,
        "provenance": (
            "Generated from the same constants the enforcement code imports, "
            "not hand-written documentation."
        ),
        "iam_note": (
            "Each worker runs as its own service account with least-privilege "
            "IAM, codified in infra/scripts/. The only identity that can change "
            "live infrastructure is the apply worker's service account, and "
            "only after an operator approves the exact plan."
        ),
        "workloads": workloads,
        "human_gates": [dict(g) for g in HUMAN_GATES],
        "denylist": {
            "summary": (
                "Before any apply, the plan is checked against a fail-closed "
                "denylist. A violation blocks the apply; operator approval "
                "cannot override it."
            ),
            "enforced_at": [
                "the trusted plan-builder CI, before a plan is ever stored",
                "the approval page, as an advisory check before you approve",
                "the tofu-apply worker, immediately before apply (final gate)",
            ],
            "rules": rules,
            "adoptable_resource_types": [
                {"type": t, "label": ADOPTABLE_TYPE_LABELS[t]}
                for t in sorted(ADOPTABLE_RESOURCE_TYPES)
            ],
        },
    }

File: frontend/src/lib/capabilities.ts
// capabilities.ts — types mirroring the GET /capabilities DTO (version 1) +
// display helpers for the CapabilityCard panel.
//
// The DTO is generated server-side from the same constants the enforcement
// code imports (TOOL_REGISTRY, WORKER_REGISTRY, ACTION_REGISTRY,
// MUTATION_TOOL_NAMES, RULE_DESCRIPTIONS) — it is never hand-written.
// These types mirror that shape exactly.

import type { TranslateFn, MessageKey } from './i18n';

/** A single tool available to a workload. */
export interface CapTool {
  name: string;
  description: string;
  /** True when the tool is in MUTATION_TOOL_NAMES: "writes OR rides a
   *  write-capable credential". The per-tool description carries the nuance
   *  for tools that are included for credential-containment only. */
  write_capable: boolean;
}

/** A worker that a workload can invoke. */
export interface CapWorker {
  name: string;
  description: string;
}

/** An action a workload can take. */
export interface CapAction {
  name: string;
  display_name: string;
  requires_approval: boolean;
}

/** A human-gated operation — the gate is the POST (not the GET form page). */
export interface CapGate {
  id: string;
  title: string;
  description: string;
  route: string;
  method: string;
}

/** A single denylist rule. */
export interface CapRule {
  id: string;
  description: string;
  /** Category from CATEGORY_ORDER: typically "control-plane" | "service-managed" | "iam" | "global-v1" | "structural".
   *  A future server may emit additional categories — callers must not drop unknown ones.
   *  Typed as string so the runtime unknown-category path is never unreachable. */
  category: string;
}

/** A workload as serialized by the server. */
export interface CapWorkload {
  name: string;
  /** Crew identity (e.g. "Anchor"). */
  display_name: string;
  /** Domain subtitle (e.g. "Cloud Run config"). */
  descriptor: string;
  description: string;
  /** True only when the workload has a LIVE autonomous trigger — derived
   *  server-side from AUTONOMOUS_TRIGGER_WORKLOADS (only `drift`/Anchor
   *  today), NOT from observation_kind. */
  autonomous: boolean;
  tools: CapTool[];
  workers: CapWorker[];
  actions: CapAction[];
}

/** The full GET /capabilities DTO (version 1). */
export interface Capabilities {
  version: 1;
  provenance: string;
  iam_note: string;
  workloads: CapWorkload[];
  human_gates: CapGate[];
  denylist: {
    summary: string;
    enforced_at: string[];
    rules: CapRule[];
    /** Optional — present from Phase-2 import-admission; absent on older payloads. */
    adoptable_resource_types?: { type: string; label: string }[];
  };
}

// ---------------------------------------------------------------------------
// Denylist category headings
// ---------------------------------------------------------------------------

/** i18n key for each known denylist category's heading. Not exported — callers
 *  go through `categoryHeading()`, which is the safe (never-throws-on-unknown)
 *  entry point. */
const CATEGORY_HEADING_KEYS: Record<string, MessageKey> = {
  'control-plane': 'capability.category.controlPlane',
  'service-managed': 'capability.category.serviceManaged',
  iam: 'capability.category.iam',
  'global-v1': 'capability.category.globalV1',
  structural: 'capability.category.structural',
};

/**
 * Human-readable heading for a denylist category. A static Record built at
 * module eval (like the old `CATEGORY_HEADINGS`) can't react to a locale
 * toggle, so this resolves at RENDER time via the passed `t` — callers must
 * invoke it reactively (`categoryHeading(group.category, $t)`).
 *
 * Unknown-category behavior (safety invariant, unchanged from the old static
 * CATEGORY_HEADINGS map): a category with no key in CATEGORY_HEADING_KEYS
 * returns the raw category string rather than throwing or dropping the rule
 * group — safety information must not silently disappear because the server
 * grew a category before the frontend learned its heading.
 */
export function categoryHeading(category: string, t: TranslateFn): string {
  const key = CATEGORY_HEADING_KEYS[category];
  return key ? t(key) : category;
}

/** A single group returned by groupRules. */
export interface RuleGroup {
  /** Category string — one of the known literals or a raw unknown string.
   *  Resolve the display heading via `categoryHeading(category, t)` at render. */
  category: string;
  rules: CapRule[];
}

/**
 * Group rules by category, preserving server order within and across groups.
 *
 * Unknown-category behavior (safety invariant): a rule whose `category` is
 * not a known category is NEVER dropped — it is appended as a trailing group.
 * `categoryHeading()` resolves its display heading to the raw category string.
 * Safety information must not silently disappear because the server grew a
 * category before the frontend learned its heading.
 */
export function groupRules(rules: CapRule[]): RuleGroup[] {
  // First pass: collect groups in insertion order (preserving server order).
  const knownCategories = Object.keys(CATEGORY_HEADING_KEYS);
  const knownMap = new Map<string, CapRule[]>();
  const unknownMap = new Map<string, CapRule[]>();

  for (const rule of rules) {
    if (knownCategories.includes(rule.category)) {
      if (!knownMap.has(rule.category)) knownMap.set(rule.category, []);
      knownMap.get(rule.category)!.push(rule);
    } else {
      if (!unknownMap.has(rule.category)) unknownMap.set(rule.category, []);
      unknownMap.get(rule.category)!.push(rule);
    }
  }

  const groups: RuleGroup[] = [];

  for (const [cat, catRules] of knownMap) {
    groups.push({ category: cat, rules: catRules });
  }

  for (const [cat, catRules] of unknownMap) {
    groups.push({ category: cat, rules: catRules });
  }

  return groups;
}

// ---------------------------------------------------------------------------
// Backend capability metadata — stable-id JA maps with EN (DTO) fallback
// ---------------------------------------------------------------------------
//
// GET /capabilities carries backend-generated English prose. Per the i18n
// design doc's "Backend capability prose" rule: localize STABLE IDENTIFIERS
// (gate.id, rule.id, tool/worker name, action name) via a frontend JA map,
// falling back to the DTO's own English string when the id isn't in the map
// (an id the map hasn't caught up to yet, or a genuinely new one — never a
// blank render, never a thrown error). Arbitrary free prose with no stable id
// (denylist.summary, denylist.enforced_at[], iam_note, provenance, the raw
// system-prompt text from /workloads/{name}/prompts) is intentionally left
// English — see CapabilityCard.svelte for those pass-through sites.
//
// Each map is `Record<id, MessageKey>` (not `Record<id, string>`) so lookups
// go through the same `t()` your static UI strings do — dev/test mode's
// "missing key throws" guard still catches a typo'd id→key mapping here, and
// JA→EN fallback for an unfinished translation is automatic. The id NOT being
// in the map (a case this file must always handle) is a different situation
// from a mapped id whose key is missing — that's guarded by the `key ?`
// check before ever calling `t()`.

const GATE_TITLE_KEYS: Record<string, MessageKey> = {
  iac_apply: 'capability.gate.iacApply.title',
  rollback: 'capability.gate.rollback.title',
};
export function gateTitle(gate: CapGate, t: TranslateFn): string {
  const key = GATE_TITLE_KEYS[gate.id];
  return key ? t(key) : gate.title;
}

const GATE_DESCRIPTION_KEYS: Record<string, MessageKey> = {
  iac_apply: 'capability.gate.iacApply.description',
  rollback: 'capability.gate.rollback.description',
};
export function gateDescription(gate: CapGate, t: TranslateFn): string {
  const key = GATE_DESCRIPTION_KEYS[gate.id];
  return key ? t(key) : gate.description;
}

// Keyed by the exact rule IDs in driftscribe_lib/iac_plan_denylist.py::RULE_DESCRIPTIONS.
const RULE_DESCRIPTION_KEYS: Record<string, MessageKey> = {
  'plan-json-unparseable': 'capability.rule.planJsonUnparseable',
  'plan-json-missing-resource-changes': 'capability.rule.planJsonMissingResourceChanges',
  'plan-json-malformed-change': 'capability.rule.planJsonMalformedChange',
  'control-plane-service': 'capability.rule.controlPlaneService',
  'control-plane-sa': 'capability.rule.controlPlaneSa',
  'control-plane-bucket': 'capability.rule.controlPlaneBucket',
  'service-managed-bucket': 'capability.rule.serviceManagedBucket',
  'service-managed-pubsub': 'capability.rule.serviceManagedPubsub',
  'control-plane-secret': 'capability.rule.controlPlaneSecret',
  'control-plane-kms': 'capability.rule.controlPlaneKms',
  'wif-config-change': 'capability.rule.wifConfigChange',
  'iam-change-forbidden-v1': 'capability.rule.iamChangeForbiddenV1',
  'import-with-changes-forbidden-v1': 'capability.rule.importWithChangesForbiddenV1',
  'import-type-not-adoptable-v1': 'capability.rule.importTypeNotAdoptableV1',
  'import-mixed-plan-forbidden-v1': 'capability.rule.importMixedPlanForbiddenV1',
  'import-batch-forbidden-v1': 'capability.rule.importBatchForbiddenV1',
  'delete-action-forbidden-v1': 'capability.rule.deleteActionForbiddenV1',
  'forget-action-forbidden-v1': 'capability.rule.forgetActionForbiddenV1',
  'replace-action-forbidden-v1': 'capability.rule.replaceActionForbiddenV1',
  'unknown-action-forbidden-v1': 'capability.rule.unknownActionForbiddenV1',
};
export function ruleDescription(rule: CapRule, t: TranslateFn): string {
  const key = RULE_DESCRIPTION_KEYS[rule.id];
  return key ? t(key) : rule.description;
}

// Keyed by the exact tool names in agent/capabilities.py::TOOL_DESCRIPTIONS.
const TOOL_DESCRIPTION_KEYS: Record<string, MessageKey> = {
  drift_read_live_env: 'capability.tool.driftReadLiveEnv',
  read_project_inventory: 'capability.tool.readProjectInventory',
  drift_patch_docs: 'capability.tool.driftPatchDocs',
  drift_propose_rollback: 'capability.tool.driftProposeRollback',
  notify: 'capability.tool.notify',
  load_contract: 'capability.tool.loadContract',
  search_recent_prs: 'capability.tool.searchRecentPrs',
  load_iac_plan: 'capability.tool.loadIacPlan',
  read_team_log: 'capability.tool.readTeamLog',
  read_conversations: 'capability.tool.readConversations',
  upgrade_read_dependencies: 'capability.tool.upgradeReadDependencies',
  upgrade_propose_pr: 'capability.tool.upgradeProposePr',
  upgrade_close_pr: 'capability.tool.upgradeClosePr',
  upgrade_merge_pr: 'capability.tool.upgradeMergePr',
  search_developer_docs: 'capability.tool.searchDeveloperDocs',
  retrieve_developer_doc: 'capability.tool.retrieveDeveloperDoc',
  provision_open_infra_pr: 'capability.tool.provisionOpenInfraPr',
  provision_propose_adoption: 'capability.tool.provisionProposeAdoption',
  get_session_state: 'capability.tool.getSessionState',
  set_session_state: 'capability.tool.setSessionState',
};
export function toolDescription(tool: CapTool, t: TranslateFn): string {
  const key = TOOL_DESCRIPTION_KEYS[tool.name];
  return key ? t(key) : tool.description;
}

// Keyed by the exact worker names in agent/capabilities.py::WORKER_DESCRIPTIONS.
const WORKER_DESCRIPTION_KEYS: Record<string, MessageKey> = {
  drift_reader: 'capability.worker.driftReader',
  drift_docs: 'capability.worker.driftDocs',
  drift_rollback: 'capability.worker.driftRollback',
  infra_reader: 'capability.worker.infraReader',
  notifier: 'capability.worker.notifier',
  upgrade_reader: 'capability.worker.upgradeReader',
  upgrade_docs: 'capability.worker.upgradeDocs',
  tofu_editor: 'capability.worker.tofuEditor',
};
export function workerDescription(worker: CapWorker, t: TranslateFn): string {
  const key = WORKER_DESCRIPTION_KEYS[worker.name];
  return key ? t(key) : worker.description;
}

// Keyed by the exact action names in agent/workloads/registry.py::ACTION_REGISTRY.
// `no_op`'s display_name ("No action needed") is byte-identical to the decision-log
// label already carried in shared.decision.noOp — reuse that key instead of a
// second copy of the same sentence.
const ACTION_DISPLAY_NAME_KEYS: Record<string, MessageKey> = {
  docs_pr: 'capability.action.docsPr',
  drift_issue: 'capability.action.driftIssue',
  escalation: 'capability.action.escalation',
  no_op: 'shared.decision.noOp',
  rollback: 'capability.action.rollback',
  upgrade_pr: 'capability.action.upgradePr',
};
export function actionDisplayName(action: CapAction, t: TranslateFn): string {
  const key = ACTION_DISPLAY_NAME_KEYS[action.name];
  return key ? t(key) : action.display_name;
}

// Keyed by the exact resource types in agent/capabilities.py::ADOPTABLE_TYPE_LABELS.
const ADOPTABLE_TYPE_LABEL_KEYS: Record<string, MessageKey> = {
  google_storage_bucket: 'capability.adoptableType.googleStorageBucket',
  google_pubsub_topic: 'capability.adoptableType.googlePubsubTopic',
  google_pubsub_subscription: 'capability.adoptableType.googlePubsubSubscription',
  google_cloud_run_v2_service: 'capability.adoptableType.googleCloudRunV2Service',
};
export function adoptableTypeLabel(entry: { type: string; label: string }, t: TranslateFn): string {
  const key = ADOPTABLE_TYPE_LABEL_KEYS[entry.type];
  return key ? t(key) : entry.label;
}
